spec/ics-4-channel-and-packet-semantics/README.md

```diff
@@ -0,0 +1,384 @@
+
+For example, an application may wish to allow a single tokenized asset to be transferred between and held on multiple blockchains while preserving fungibility and conservation of supply. The application can mint asset vouchers on chain `B` when a particular IBC packet is committed to chain `B`, and require outgoing sends of that packet on chain `A` to escrow an equal amount of the asset on chain `A` until the vouchers are later redeemed back to chain `A` with an IBC packet in the reverse direction. This ordering guarantee along with correct application logic can ensure that total supply is preserved across both chains and that any vouchers minted on chain `B` can later be redeemed back to chain `A`.
+
+#### Permissioning
+
+- Channel ends should be permissioned to one module on each end.
+
+## Technical Specification
+
+![channel-state-machine](channel-state-machine.png)
+
+### Channel opening handshake
+
+```typescript
+interface ChanOpenInit {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  counterpartyChannelIdentifier: Identifier
+  counterpartyModuleIdentifier: Identifier
+}
+```
+
+```typescript
+function chanOpenInit() {
+  moduleIdentifier = getCallingModule()
+  assert(get("connections/{connectionIdentifier}/channels/{channelIdentifier}") === null)
+  connection = get("connections/{connectionIdentifier}")
+  assert(connection.state === OPEN)
+  channel = Channel{INIT, moduleIdentifier, counterpartyModuleIdentifier, counterpartyChannelIdentifier, 0, 0}
+  set("connections/{connectionIdentifier}/channels/{channelIdentifier}", channel)
+}
+```
+
+```typescript
+interface ChanOpenTry {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  counterpartyChannelIdentifier: Identifier
+  moduleIdentifier: Identifier
+  counterpartyModuleIdentifier: Identifier
+  proofInit: CommitmentProof
+}
+```
+
+```typescript
+function chanOpenTry() {
+  assert(get("connections/{connectionIdentifier}/channels/{channelIdentifier}") === null)
+  assert(getCallingModule() === moduleIdentifier)
+  connection = get("connections/{connectionIdentifier}")
+  assert(connection.state === OPEN)
+  consensusState = get("clients/{connection.clientIdentifier}")
+  assert(verifyMembership(
+    consensusState,
+    proofInit,
+    "connections/{connection.counterpartyConnectionIdentifier}/channels/{counterpartyChannelIdentifier}",
+    Channel{INIT, counterpartyModuleIdentifier, moduleIdentifier, channelIdentifier, 0, 0}
+  ))
+  channel = Channel{TRYOPEN, moduleIdentifier, counterpartyModuleIdentifier, counterpartyChannelIdentifier, 0, 0}
+  set("connections/{connectionIdentifier}/channels/{channelIdentifier}", channel)
+}
+```
+
+```typescript
+interface ChanOpenAck {
+  connectionIdentifier: Identifier
+  channelIdentifier: Identifier
+  proofTry: CommitmentProof
+}
+```
```

Comment:
At this point in the reading I realize that I don't understand the scope of the channelIdentifier. Is it unique per module, with different modules being able to use the same identifier? In which case we would need to keep carrying in all these interfaces the moduleIdentifier.
IMO we need a separate section that covers this, how these identifiers are allocated, their scope, etc. If this is out of the scope of this specification we can just say so and point to another ICS.

Reply:
Channel identifiers are unique per connection. Each channel end is owned by a particular module.

Working on a diagram; added a paragraph on channel ownership at the top of this section.